archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,261 @@
+      suffices : sqrt (1 + n) - 1 ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+      { linarith },
+      have t :  (⌈sqrt (1 + n) - 1⌉:ℝ) ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+      { norm_cast,
+        exact ceil_le_floor_add_one _ },
+      apply le_trans _ t,
+      exact le_ceil (sqrt (1 + n) - 1) },
+      suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+      { linarith },
+    exact radical_inequality hn },
+  { rw hx,
+    exact floor_le _ },
+  { rw [floor_nonneg, le_sub, sub_zero, le_sqrt],
+    all_goals { norm_cast, linarith } },
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : 100 ≤ n):
+  (∃ (a b c : ℕ), a ∈ finset.Ico n (2 * n + 1) ∧ b ∈ finset.Ico n (2 * n + 1) ∧
+  c ∈ finset.Ico n (2 * n + 1) ∧ (a : ℕ) < b ∧ (b : ℕ) < c ∧ (∃ (k : ℕ), (a : ℕ) + b = k * k) ∧
+  (∃ (l : ℕ), (c : ℕ) + a = l * l) ∧ (∃ (m : ℕ), (b : ℕ) + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : 107 ≤ n,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    by_cases p : 1 < l,
+    { have h₁ : 4 * l ≤ 2 * l * l, { linarith },
+      have h₂ :  2 * l * l - 4 * l ∈ finset.Ico n (2 * n + 1),
+      { rw finset.Ico.mem,
+        zify [h₁],
+        split,
+        { linarith },
+        { linarith }},
+      have h₃ : 2 * l * l + 4 * l ∈ finset.Ico n (2 * n + 1),
+      { rw finset.Ico.mem,
+        split,
+        { linarith },
+        { linarith }},
+      have h₄ : 2 * l * l - 4 * l < 2 * l * l + 1,
+      { zify [h₁],
+        linarith },
+      have h₅ : 2 * l * l + 1 < 2 * l * l + 4 * l, { linarith },
+      have h₆ : 2 * l * l + 1 ∈ finset.Ico n (2 * n + 1),
+      { rw finset.Ico.mem,
+        split,
+        { rw finset.Ico.mem at h₂,
+          linarith },
+        { rw finset.Ico.mem at h₃,
+          linarith }},
+      use [2 * l * l - 4 * l, 2 * l * l + 1, 2 * l * l + 4 * l],
+      refine ⟨h₂,⟨h₆,⟨h₃,⟨h₄,⟨h₅,⟨_,⟨_,_⟩⟩⟩⟩⟩⟩⟩,
+      { use (2 * l - 1),
+        have h₇ : 1 ≤ 2 * l, { linarith },
+        zify [h₁, h₇],
+        ring_exp_eq },
+      { use (2 * l),
+        zify [h₁],
+        ring_exp_eq },
+      { use(2 * l + 1),
+        ring_exp_eq }},
+    { exfalso,
+      simp only [not_lt] at p,
+      interval_cases l,
+      all_goals {linarith } }},
+  --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+  --construction of a triplet {a, b, c}, which is constructed by setting l=9
+  --in the argument at the start of the file.
+  { have h₈ : 126 ∈ finset.Ico n (2 * n + 1),
+    { rw finset.Ico.mem,
+      split,
+      all_goals { linarith } },
+      have h₉ : 163 ∈ finset.Ico n (2 * n + 1),
+    { rw finset.Ico.mem,
+      split,
+      all_goals { linarith } },
+    have h₁₀ : 198 ∈ finset.Ico n (2 * n + 1),
+    { rw finset.Ico.mem,
+      split,
+      all_goals { linarith } },
+    use [126, 163, 198],
+    refine ⟨h₈,⟨h₉,⟨h₁₀,⟨_,⟨_,⟨⟨17, _⟩,⟨⟨18, _⟩, ⟨19, _⟩⟩⟩⟩⟩⟩⟩⟩,
+    all_goals { linarith } },
```

Comment:
```suggestion
    refine ⟨126, 163, 198, h₈, h₉, h₁₀, _, _, ⟨17, _⟩, ⟨18, _⟩, 19, _⟩; linarith }
```
and you can replace all your `something, all_goals { linarith }` by `something; linarith`. The `;` makes the next tactic apply to all goals.

Reply:
> and you can replace all your `something, all_goals { linarith }` by `something; linarith`. The `;` makes the next tactic apply to all goals.

Thanks for telling me this! However, is the `;` to be used instead of `all_goals` also, when `all_goals` takes in multiple tactics?

Reply:
Regarding this, I'm seeing behaviour, where sometimes replacing `all_goals { norm_cast, linarith }` by just 
`{ norm_cast, linarith }` with `;` before doesn't close the goal.

Reply:
You can chain multiple `;`. So `something, all_goals { norm_cast, linarith }` becomes `something; norm_cast; linarith`. Note that you can also use `;` after a `by` without needing curly braces but this use is **deprecated** (you can read about it in the style guide).

Reply:
Another alternative, which I think is clearer, would be `something; { norm_cast, linarith }`. I really prefer if `;` is *only* used at the point multiple goals are created, so that the "goal flow" is visible from the syntax.